### PR TITLE
MM-32882: Add an unread badge/marker to the Main Menu icon and the ‘Plugin Marketplace’ Menu Item

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Add an unread badge/marker to the Main Menu icon and the ‘Plugin Marketplace’ Menu Item until an admin user visits the marketplace for the first time

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32882

Related PRs
server: https://github.com/mattermost/mattermost-server/pull/16992
webapp: https://github.com/mattermost/mattermost-webapp/pull/7563

### CODE REVIEW

File: v4/source/system.yaml [hunk 1]
```diff
@@ -1488,3 +1488,54 @@
...
+        Users must be system admins - have `manage_system` permissions
+      responses:
+        "200":
+          description: Retrieves the system-level status
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/System"
+        "400":
+          $ref: "#/components/responses/BadRequest"
```
Comment (hanzei) [inline v4/source/system.yaml:1509]: Is there a case were the code throws this error?

File: v4/source/system.yaml [hunk 2]
```diff
@@ -1488,3 +1488,54 @@
...
+        required: true
+      responses:
+        "200":
+          description: setting has been successfully set
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/StatusOK"
+        "400":
+          $ref: "#/components/responses/BadRequest"
```
Comment (hanzei) [inline v4/source/system.yaml:1537]: Same question here

**Review by hanzei — CHANGES_REQUESTED** [review]
Two questions to double check, otherwise LGTM

**Comment by larkox** [comment]
Shouldn't @justinegeffen be assigned as reviewer here?

**Review by larkox — APPROVED** [review]
LGTM

File: v4/source/system.yaml [hunk 3]
```diff
@@ -1488,3 +1488,54 @@
...
+  /system/first_admin_visit_marketplace:
+    get:
+      tags:
+        - system
+      summary: Retrieves the system-level status that specifies that at least an admin has visited the in-product Plugin Marketplace.
+      description: >
+        Retrieves the status that specifies that at least an admin has visited the in-product Plugin Marketplace.
+        __Minimum server version: 5.33__
+        ##### Permissions
+        Users must be system admins - have `manage_system` permissions
```
Comment (justinegeffen) [inline v4/source/system.yaml:1500]:
```suggestion
        Users must be System Admins - have `manage_system` permissions.
```

File: v4/source/system.yaml [hunk 4]
```diff
@@ -1488,3 +1488,54 @@
           $ref: "#/components/responses/Forbidden"
         "404":
           $ref: "#/components/responses/NotFound"
+  /system/first_admin_visit_marketplace:
+    get:
+      tags:
+        - system
+      summary: Retrieves the system-level status that specifies that at least an admin has visited the in-product Plugin Marketplace.
```
Comment (justinegeffen) [inline v4/source/system.yaml:1495]:
```suggestion
      summary: Retrieves the system-level status that specifies that at least an Admin has visited the in-product Plugin Marketplace.
```

File: v4/source/system.yaml [hunk 5]
```diff
@@ -1488,3 +1488,54 @@
...
+        "400":
+          $ref: "#/components/responses/BadRequest"
+        "403":
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+    post:
+      tags:
+        - system
+      summary: Stores the system-level status that specifies that at least an admin has visited the in-product Plugin Marketplace.
```
Comment (justinegeffen) [inline v4/source/system.yaml:1517]:
```suggestion
      summary: Stores the system-level status that specifies that at least an Admin has visited the in-product Plugin Marketplace.
```

File: v4/source/system.yaml [hunk 6]
```diff
@@ -1488,3 +1488,54 @@
...
+      description: >
+        Stores the system-level status that specifies that at least an admin has visited the in-product Plugin Marketplace.
+        __Minimum server version: 5.33__
+        ##### Permissions
+        Users must be system admins - have `manage_system` permissions
```
Comment (justinegeffen) [inline v4/source/system.yaml:1522]:
```suggestion
        Users must be System Admins - have `manage_system` permissions.
```

**Review by justinegeffen — APPROVED** [review]
Thanks, @larkox! Some minor suggestions but nothing blocking.

File: v4/source/system.yaml (hunk 4 above)
Comment (justinegeffen) [reply v4/source/system.yaml:1495]: One question: When we say "at least an Admin" do we mean "at the very least, an Admin visited the Marketplace" or "at least one Admin visited..." or something else? The sentence is a little unclear in that respect.

Comment (catalintomai) [reply v4/source/system.yaml:1495]: @justinegeffen - "at least one Admin visited..." - or in other words, the first one to visit the in-product Plugin Marketplace turns the badge off.

Comment (justinegeffen) [reply v4/source/system.yaml:1495]: Perfect, thank you!

**Review by jfrerich — APPROVED** [review]

**Review by hanzei — APPROVED** [review]

File: v4/source/system.yaml [hunk 7]
```diff
@@ -1488,3 +1488,50 @@
           $ref: "#/components/responses/Forbidden"
         "404":
           $ref: "#/components/responses/NotFound"
+  /system/first_admin_visit_marketplace:
```
Comment (hanzei) [inline v4/source/system.yaml:1491]: This need to get updated

**Review by hanzei — CHANGES_REQUESTED** [review]

File: v4/source/plugins.yaml [hunk 8]
```diff
@@ -553,3 +553,50 @@
             }
 
             ok, resp = Client.GetMarketplacePlugins(filter)
+  /plugins/first_admin_visit_marketplace:
+    get:
+      tags:
+        - system
```
Comment (hanzei) [inline v4/source/plugins.yaml:559]: Should that be
```suggestion
        - plugins
```
?

File: v4/source/plugins.yaml [hunk 9]
```diff
@@ -553,3 +553,50 @@
...
+      summary: Retrieves the system-level status that specifies that at least one System Admin has visited the in-product Plugin Marketplace.
```
Comment (hanzei) [inline v4/source/plugins.yaml:560]: The summary is quite long and clutters the side bar. Can we shorten it a bit to something like 
```suggestion
      summary: Get if the Plugin Marketplace has been visited`
```

File: v4/source/plugins.yaml [hunk 10]
```diff
@@ -553,3 +553,50 @@
...
+      description: |
+        Retrieves the status that specifies that at least one System Admin has visited the in-product Plugin Marketplace.
+        __Minimum server version: 5.33__
+        ##### Permissions
+        Users must be System Admins - have `manage_system` permissions.
```
Comment (hanzei) [inline v4/source/plugins.yaml:565]: At least in the plugin section this wording is used. Should we stick to it for consistency? 
```suggestion
        Must have `manage_system` permissions.
```

**Review by hanzei — CHANGES_REQUESTED** [review]

**Review by hanzei — APPROVED** [review]
LGTM :+1:

File: v4/source/plugins.yaml [hunk 11]
```diff
@@ -553,3 +553,50 @@
...
 
             ok, resp = Client.GetMarketplacePlugins(filter)
+  /plugins/marketplace/first_admin_visit:
+    get:
+      tags:
+        - plugins
+      summary: Get if the Plugin Marketplace has been visited by at least an admin.
+      description: |
+        Retrieves the status that specifies that at least one System Admin has visited the in-product Plugin Marketplace.
+        __Minimum server version: 5.33__
```
Comment (hanzei) [inline v4/source/plugins.yaml:563]: Does this need to get updated?

File: v4/source/plugins.yaml [hunk 12]
```diff
@@ -553,3 +553,50 @@
...
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+    post:
+      tags:
+        - system
+      summary: Stores that the Plugin Marketplace has been visited by at least an admin.
+      description: |
+        Stores the system-level status that specifies that at least an admin has visited the in-product Plugin Marketplace.
+        __Minimum server version: 5.33__
```
Comment (hanzei) [inline v4/source/plugins.yaml:583]: Same question here

**Pull request merged**